core/src/repair_generic_traversal.rs

```diff
@@ -0,0 +1,229 @@
+    unknown_last
+        .iter()
+        .take(limit)
+        .map(|(slot, received, _)| ShredRepairType::HighestShred(*slot, *received))
+        .collect()
+}
+
+fn get_unrepaired_path(
+    start_slot: Slot,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    visited: &mut HashSet<Slot>,
+) -> Vec<Slot> {
+    let mut path = Vec::new();
+    let mut slot = start_slot;
+    while !visited.contains(&slot) {
+        visited.insert(slot);
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.is_full() {
+                break;
+            }
+            path.push(slot);
+            slot = slot_meta.parent_slot;
+        }
+    }
+    path.reverse();
+    path
+}
+
+pub fn get_closest_completion(
+    tree: &HeaviestSubtreeForkChoice,
+    blockstore: &Blockstore,
+    slot_meta_cache: &mut HashMap<Slot, Option<SlotMeta>>,
+    processed_slots: &mut HashSet<Slot>,
+    limit: usize,
+) -> Vec<ShredRepairType> {
+    let mut v: Vec<(Slot, u64)> = Vec::default();
+    let iter = GenericTraversal::new(tree);
+    for slot in iter {
+        if processed_slots.contains(&slot) {
+            continue;
+        }
+        let slot_meta = slot_meta_cache
+            .entry(slot)
+            .or_insert_with(|| blockstore.meta(slot).unwrap());
+        if let Some(slot_meta) = slot_meta {
+            if slot_meta.is_full() {
+                continue;
+            }
+            if let Some(last_index) = slot_meta.known_last_index() {
+                let shred_index = blockstore.get_index(slot).unwrap();
+                let dist = if let Some(shred_index) = shred_index {
+                    let shred_count = shred_index.data().num_shreds() as u64;
+                    last_index - shred_count
+                } else {
+                    last_index - slot_meta.consumed
+                };
+                v.push((slot, dist));
+                processed_slots.insert(slot);
+            }
+        }
+    }
+    v.sort_by(|(_, d1), (_, d2)| d1.cmp(d2));
+
+    let mut visited = HashSet::new();
+    let mut repairs = Vec::new();
+    for (slot, _) in v {
+        if repairs.len() >= limit {
+            break;
+        }
+        // attempt to repair heaviest slots starting with their parents
+        let path = get_unrepaired_path(slot, blockstore, slot_meta_cache, &mut visited);
+        for slot in path {
+            if repairs.len() >= limit {
+                break;
+            }
+            let slot_meta = slot_meta_cache.get(&slot).unwrap().as_ref().unwrap();
+            let new_repairs = RepairService::generate_repairs_for_slot(
+                blockstore,
+                slot,
+                slot_meta,
+                limit - repairs.len(),
+            );
+            repairs.extend(new_repairs);
+        }
+    }
```

Comment:
maybe a unit test for this case to make sure parents are being repaired in order